Implement ToggleEvent.source

lukewarlow opened the pull request:
#### c4ee30d3b1b98692df765d43c8b07dd3f1674de2
<pre>
Implement ToggleEvent.source
<a href="https://bugs.webkit.org/show_bug.cgi?id=293686">https://bugs.webkit.org/show_bug.cgi?id=293686</a>

Reviewed by Tim Nguyen and Anne van Kesteren.

This implements the new source attribute on ToggleEvent, and wires it up from command invokers for dialogs,
and all invokers for popover.

This also tweaks the source property getter on CommandEvent.

* LayoutTests/imported/w3c/web-platform-tests/html/dom/idlharness.https_exclude=(Document_Window_HTML._)-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/details-toggle-source-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-dialog-element/dialog-toggle-source-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-toggle-source-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/the-button-element/command-and-commandfor/source-attribute-retargeting.tentative-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/dom/idlharness.https_exclude=(Document_Window_HTML._)-expected.txt:
* Source/WebCore/dom/CommandEvent.cpp:
(WebCore::CommandEvent::source const):
* Source/WebCore/dom/ToggleEvent.cpp:
(WebCore::ToggleEvent::ToggleEvent):
(WebCore::ToggleEvent::source const):
* Source/WebCore/dom/ToggleEvent.h:
* Source/WebCore/dom/ToggleEvent.idl:
* Source/WebCore/dom/ToggleEventTask.cpp:
(WebCore::ToggleEventTask::queue):
* Source/WebCore/dom/ToggleEventTask.h:
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::queueDetailsToggleEventTask):
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::show):
(WebCore::HTMLDialogElement::showModal):
(WebCore::HTMLDialogElement::close):
(WebCore::HTMLDialogElement::requestClose):
(WebCore::HTMLDialogElement::handleCommandInternal):
(WebCore::HTMLDialogElement::queueDialogToggleEventTask):
* Source/WebCore/html/HTMLDialogElement.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::queuePopoverToggleEventTask):
(WebCore::HTMLElement::showPopoverInternal):
(WebCore::HTMLElement::hidePopoverInternal):
(WebCore::HTMLElement::togglePopover):
(WebCore::HTMLElement::handleCommandInternal):
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::handlePopoverTargetAction):

Canonical link: <a href="https://commits.webkit.org/306152@main">https://commits.webkit.org/306152@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/759b9ecddf4140b7c884bfbd5795549d147fe133

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [⏳ 🧪 style ](https://ews-build.webkit.org/#/builders/Style-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-26-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac-debug ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe-cairo-libwebrtc ](https://ews-build.webkit.org/#/builders/WPE-Cairo-LibWebRTC-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc-debug-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-O3-Debug-arm64-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-26-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/29585 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 playstation ](https://ews-build.webkit.org/#/builders/PlayStation-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-26-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-intel-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-26-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-26-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-26-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/ToggleEventTask.cpp:
```diff
@@ -56,7 +56,11 @@ void ToggleEventTask::queue(ToggleState oldState, ToggleState newState)
 
         auto data = *std::exchange(task->m_data, std::nullopt);
-        element->dispatchEvent(ToggleEvent::create(eventNames().toggleEvent, { EventInit { }, stringForState(data.oldState), stringForState(data.newState) }, Event::IsCancelable::No));
+        ToggleEvent::Init init;
+        init.oldState = stringForState(data.oldState);
+        init.newState = stringForState(data.newState);
+        init.source = data.source;
```
```suggestion
        ToggleEvent::Init init {
            .oldState = stringForState(data.oldState),
            .newState = stringForState(data.newState),
            .source = data.source
        };
```

lukewarlow replied:
Based on the build failures on GTK and WPE it looks like this pattern might not work?

I'm guessing it doesn't like that EventInit properties aren't being explicitly initialized? Not entirely sure.

In file included from WebCore/DerivedSources/unified-sources/UnifiedSource-be65d27a-25.cpp:2:
../../../Source/WebCore/dom/ToggleEventTask.cpp: In lambda function:
../../../Source/WebCore/dom/ToggleEventTask.cpp:63:9: error: missing initializer for member ‘WebCore::ToggleEvent::Init::<anonymous>’ [-Werror=missing-field-initializers]
   63 |         };
      |         ^
cc1plus: all warnings being treated as errors

nt1m commented on Source/WebCore/html/HTMLDialogElement.cpp:
```diff
@@ -111,7 +115,11 @@ ExceptionOr<void> HTMLDialogElement::showModal()
         return Exception { ExceptionCode::InvalidStateError, "Invalid for dialogs within documents that are not fully active."_s };
 
-    Ref event = ToggleEvent::create(eventNames().beforetoggleEvent, { EventInit { }, "closed"_s, "open"_s }, Event::IsCancelable::Yes);
+    ToggleEvent::Init init;
+    init.oldState = "closed"_s;
+    init.newState = "open"_s;
+    init.source = source;
```
ditto

nt1m commented on Source/WebCore/html/HTMLDialogElement.h:
```diff
@@ -55,7 +55,7 @@ class HTMLDialogElement final : public HTMLElement {
     bool handleCommandInternal(HTMLButtonElement& invoker, const CommandType&) final;
 
-    void queueDialogToggleEventTask(ToggleState oldState, ToggleState newState);
+    void queueDialogToggleEventTask(ToggleState oldState, ToggleState newState, RefPtr<Element> source);
```
This should take a `Element*`

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1131,15 +1131,19 @@ ExceptionOr<void> HTMLElement::showPopoverInternal(HTMLElement* invoker)
 
     Ref document = this->document();
-    Ref event = ToggleEvent::create(eventNames().beforetoggleEvent, { EventInit { }, "closed"_s, "open"_s }, Event::IsCancelable::Yes);
+    ToggleEvent::Init init;
+    init.oldState = "closed"_s;
+    init.newState = "open"_s;
+    init.source = source;
```
ditto

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1230,8 +1234,14 @@ ExceptionOr<void> HTMLElement::hidePopoverInternal(FocusPreviousElement focusPre
-        dispatchEvent(ToggleEvent::create(eventNames().beforetoggleEvent, { EventInit { }, "open"_s, "closed"_s }, Event::IsCancelable::No));
+    if (fireEvents == FireEvents::Yes) {
+        ToggleEvent::Init init;
+        init.source = source;
+        init.oldState = "open"_s;
+        init.newState = "closed"_s;
```
ditto

annevk commented on Source/WebCore/dom/ToggleEvent.cpp:
```diff
@@ -66,4 +68,20 @@ Ref<ToggleEvent> ToggleEvent::createForBindings()
+    if (RefPtr target = dynamicDowncast<Node>(currentTarget())) {
+        Ref treeScope = target->treeScope();
+        Ref node = treeScope->retargetToScope(*m_source.get());
+        return &downcast<Element>(node).get();
```
I think this should work:
```suggestion
        return downcast<Element>(node);
```
And if so, please also apply below.

lukewarlow replied:
error: taking the address of a temporary object of type 'Ref<match_constness_t<Node, Element>>'

Looks like it needs the .get() it's a Ref<Eelement> * without it.

annevk replied:
Okay, in that case `downcast<Element>(node).ptr()` would be slightly better I suppose, but it still seems weird to me. Also, does that mean this can return `Element*` instead of `RefPtr<Element>`? That's generally preferable for non-allocating methods.

lukewarlow replied:
Managed to get return downcast<Element>(node); working, not sure what I did wrong the first time.

annevk commented on Source/WebCore/dom/ToggleEventTask.cpp:
```diff
@@ -37,7 +37,7 @@ Ref<ToggleEventTask> ToggleEventTask::create(Element& element)
 }
 
-void ToggleEventTask::queue(ToggleState oldState, ToggleState newState)
+void ToggleEventTask::queue(ToggleState oldState, ToggleState newState, RefPtr<Element> source)
```
RefPtr isn't needed here. This can be Element*.

annevk commented on Source/WebCore/html/HTMLDialogElement.cpp:
```diff
@@ -292,12 +304,12 @@ void HTMLDialogElement::setIsModal(bool newValue)
 }
 
-void HTMLDialogElement::queueDialogToggleEventTask(ToggleState oldState, ToggleState newState)
+void HTMLDialogElement::queueDialogToggleEventTask(ToggleState oldState, ToggleState newState, RefPtr<Element> source)
```
Don't think this needs a RefPtr.

annevk commented on Source/WebCore/html/HTMLDialogElement.h:
```diff
@@ -42,9 +42,9 @@ class HTMLDialogElement final : public HTMLElement {
-    ExceptionOr<void> showModal();
-    void close(const String&);
-    void requestClose(const String&);
+    ExceptionOr<void> showModal(Element* = nullptr);
+    void close(const String&, Element* = nullptr);
+    void requestClose(const String&, Element* = nullptr);
```
If there's not too many callers it might be nicer if they're all explicit?

lukewarlow replied:
These are called directly by WebIDL so they require a default for the second argument. Alternatively we could split them so there's a binding entry point and then the underlying one that gets called?

annevk replied:
No in that case this is fine.

annevk commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1112,17 +1112,17 @@ static void runPopoverFocusingSteps(HTMLElement& popover)
 }
 
-void HTMLElement::queuePopoverToggleEventTask(ToggleState oldState, ToggleState newState)
+void HTMLElement::queuePopoverToggleEventTask(ToggleState oldState, ToggleState newState, RefPtr<Element> source)
```
No need for RefPtr.

annevk reviewed (commented):
Thanks for working on this!

annevk commented on Source/WebCore/dom/ToggleEvent.cpp:
```diff
@@ -66,4 +72,20 @@ Ref<ToggleEvent> ToggleEvent::createForBindings()
+
+    if (RefPtr target = dynamicDowncast<Node>(currentTarget())) {
+        Ref treeScope = target->treeScope();
+        Ref node = treeScope->retargetToScope(*m_source.get());
```
We should not need `.get()` here. (Also applies below.)

annevk commented on Source/WebCore/dom/ToggleEvent.h:
```diff
@@ -54,6 +58,7 @@ class ToggleEvent final : public Event {
 
     String m_oldState;
     String m_newState;
+    RefPtr<Element> m_source;
```
```suggestion
    const RefPtr<Element> m_source;
```

annevk reviewed (approved):
Looks good modulo some minor nits.

annevk commented on Source/WebCore/dom/ToggleEvent.idl:
```diff
@@ -28,11 +28,13 @@ interface ToggleEvent : Event {
 };
 
 [
     LegacyNativeDictionaryRequiredInterfaceNullability,
```
The moment we remove this we'll have to redo all the `ToggleEvent::Init` initialization below to be complete (including the super struct) and done at once. You'll be in a race with @weinig, but he might also appreciate it if you took care of it.

nt1m replied:
Seems ok to do this a follow up, since this is pre-existing.

lukewarlow replied:
My editor has decided to stop working ATM so I vote for a follow up else it'll have to wait till I fix it 😅

lukewarlow replied:
My editor has decided to stop working ATM so I vote for a follow up else it'll have to wait till I fix it 😅

lukewarlow committed:
Implement ToggleEvent.source
https://bugs.webkit.org/show_bug.cgi?id=293686

Reviewed by Tim Nguyen and Anne van Kesteren.

This implements the new source attribute on ToggleEvent, and wires it up from command invokers for dialogs,
and all invokers for popover.

This also tweaks the source property getter on CommandEvent.

* LayoutTests/imported/w3c/web-platform-tests/html/dom/idlharness.https_exclude=(Document_Window_HTML._)-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/details-toggle-source-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-dialog-element/dialog-toggle-source-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-toggle-source-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/the-button-element/command-and-commandfor/source-attribute-retargeting.tentative-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/dom/idlharness.https_exclude=(Document_Window_HTML._)-expected.txt:
* Source/WebCore/dom/CommandEvent.cpp:
(WebCore::CommandEvent::source const):
* Source/WebCore/dom/ToggleEvent.cpp:
(WebCore::ToggleEvent::ToggleEvent):
(WebCore::ToggleEvent::source const):
* Source/WebCore/dom/ToggleEvent.h:
* Source/WebCore/dom/ToggleEvent.idl:
* Source/WebCore/dom/ToggleEventTask.cpp:
(WebCore::ToggleEventTask::queue):
* Source/WebCore/dom/ToggleEventTask.h:
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::queueDetailsToggleEventTask):
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::show):
(WebCore::HTMLDialogElement::showModal):
(WebCore::HTMLDialogElement::close):
(WebCore::HTMLDialogElement::requestClose):
(WebCore::HTMLDialogElement::handleCommandInternal):
(WebCore::HTMLDialogElement::queueDialogToggleEventTask):
* Source/WebCore/html/HTMLDialogElement.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::queuePopoverToggleEventTask):
(WebCore::HTMLElement::showPopoverInternal):
(WebCore::HTMLElement::hidePopoverInternal):
(WebCore::HTMLElement::togglePopover):
(WebCore::HTMLElement::handleCommandInternal):
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::handlePopoverTargetAction):

Canonical link: https://commits.webkit.org/306152@main
